File: ardrive_network/lib/src/ardrive_network.dart
import 'dart:async';
import 'dart:collection';
import 'dart:convert';

import 'package:ardrive_network/src/responses.dart';
import 'package:ardrive_network/src/utils.dart';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:isolated_worker/js_isolated_worker.dart';

class ArdriveNetwork {
  late int retries;
  late int retryDelayMs;
  late bool noLogs;
  bool _areScriptsImported = false;
  int retryAttempts = 0;

  ArdriveNetwork({
    int retries = 8,
    int retryDelayMs = 200,
    noLogs = false,
  }) {
    this.retries = retries;
    this.retryDelayMs = retryDelayMs;
    this.noLogs = noLogs;
  }

  Dio dio() {
    final dio = Dio();

    if (!noLogs) {
      dio.interceptors.add(LogInterceptor());
    }

    if (retries > 0) {
      final dioRetry = getDioRetrySettings(
        dio: dio,
        retries: retries,
        retryDelayMs: retryDelayMs,
        retryAttempts: retryAttempts,
        noLogs: noLogs,
      );

      dio.interceptors.add(dioRetry);
    }

    return dio;
  }

  Future<ArDriveNetworkResponse> getJson(String url) async {
    if (kIsWeb) {
      if (await _loadWebWorkers()) {
        return await _getJsonWeb(url);
      }
    }

    return await compute(_getJsonIO, url);
  }

  Future<ArDriveNetworkResponse> _getJsonIO(String url) async {
    try {
      Response<String> response = await this.dio().get(url);

      return ArDriveNetworkResponse(
          data: jsonDecode(response.data ?? ''),
          statusCode: response.statusCode,
          statusMessage: response.statusMessage,
          retryAttempts: this.retryAttempts);
    } catch (error) {
      throw ArDriveNetworkException(
          retryAttempts: this.retryAttempts, dioException: error);
    }
  }

  Future<ArDriveNetworkResponse> _getJsonWeb(String url) async {
    try {
      final LinkedHashMap<dynamic, dynamic> response =
          await JsIsolatedWorker().run(
        functionName: 'getJson',
        arguments: [
          url,
          this.retries,
          this.retryDelayMs,
          this.noLogs,
        ],
      );

      if (response['error'] != null) {
        this.retryAttempts = response['retryAttempts'];

        throw response['error'];
      }

      return ArDriveNetworkResponse(
        data: response['data'],
        statusCode: response['statusCode'],
        statusMessage: response['statusMessage'],
        retryAttempts: response['retryAttempts'],
      );
    } catch (error) {
      throw ArDriveNetworkException(
          retryAttempts: this.retryAttempts, dioException: error);
    }
  }

  Future<bool> _loadWebWorkers() async {
    const List<String> _jsScript = <String>['ardrive-network.js'];
    bool jsLoaded = false;

    if (!_areScriptsImported) {
      jsLoaded = await JsIsolatedWorker().importScripts(_jsScript);
      _areScriptsImported = !jsLoaded;
    }

    if (jsLoaded) {
      return true;
    } else {
      throw Exception('ArDriveNetwork Web worker is not available!');
    }
  }
}

File: ardrive_network/lib/src/js/ardrive-network.js
const retryStatusCodes = [
  408, 429, 440, 460, 499, 500, 502, 503, 504, 520, 521, 522, 523, 524, 525,
  527, 598, 599,
];

function retryDelay(attempt, retryDelayMS) {
  return parseInt(retryDelayMS * Math.pow(1.5, attempt));
}

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function retry(url, retries, retryDelayMs, noLogs, retryAttempts) {
  await delay(retryDelay(retryAttempts, retryDelayMs));

  return await getJson([url, retries - 1, retryDelayMs, noLogs, retryAttempts]);
}

function formatLog(url, statusCode, statusMessage, retryAttempts) {
  return `uri: ${url}
  response: Http status error [${statusCode}]: ${statusMessage}
  retryAttempts: ${retryAttempts}`;
}

function isStatusCodeError(code) {
  return code >= 400 && code <= 599;
}

async function getJson(params) {
  const url = params[0];
  const retries = params[1];
  const retryDelayMs = params[2];
  const noLogs = params[3];
  let retryAttempts = params[4] ?? 0;

  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        Accept: 'application/json',
      },
    });

    const statusCode = response.status;

    if (retries > 0 && retryStatusCodes.includes(statusCode)) {
      const log = formatLog(
        url,
        response.status,
        response.statusText,
        retryAttempts,
      );

      if (!noLogs) {
        console.warn(`Network Request Retry\n${log}`);
      }

      return await retry(url, retries, retryDelayMs, noLogs, retryAttempts + 1);
    } else {
      if (isStatusCodeError(statusCode)) {
        const log = formatLog(
          url,
          response.status,
          response.statusText,
          retryAttempts,
        );

        return {
          error: `Network Request Error\n${log}`,
          retryAttempts,
        };
      }
    }

    // if (retries === 0 || (retries === 0 && isStatusCodeError(statusCode))) {
    //   const log = formatLog(
    //     url,
    //     response.status,
    //     response.statusText,
    //     retryAttempts,
    //   );

    //   return {
    //     error: `Network Request Error\n${log}`,
    //     retryAttempts,
    //   };
    // }

    const data = await response.json();

    return {
      statusCode: response.status,
      statusMessage: response.statusText,
      data,
      retryAttempts,
    };
  } catch (error) {
    return { error, retryAttempts };
  }
}
